In [6]:
#import the necessary packages
import pandas as pd
import numpy as np
import requests

As im unable to use folium in my personal machine, I'll be completing the third section in Skills Network Labs. I have exported the df into a .csv file and have imported the same into this notebook.

In [16]:

df = pd.read_csv('df_exp.csv')
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [67]:
#there are 99 distinct neighborhoods in toronto
len(df['Neighborhood'].unique())

99

In [17]:
!conda install -c conda-forge geocoder --yes
import geocoder
from geopy.geocoders import Nominatim 

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

Solving environment: done

# All requested packages already installed.

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [20]:
import folium

# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_Toronto)  
    
map_Toronto

#### Define Foursquare Credentials and Version

In [21]:
CLIENT_ID = 'I5IGOLJKRUIUA1E4PMYEN40ZAO15GIVS33AUZS1HLDTJWLYS' # your Foursquare ID
CLIENT_SECRET = 'DFPEVUHB31554JZI5QDDRVGU5531ANMTCZSEDNOYTM51AZHG' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: I5IGOLJKRUIUA1E4PMYEN40ZAO15GIVS33AUZS1HLDTJWLYS
CLIENT_SECRET:DFPEVUHB31554JZI5QDDRVGU5531ANMTCZSEDNOYTM51AZHG


In [44]:
df.loc[99, 'Neighborhood']

'Church and Wellesley'

In [45]:
neighborhood_latitude = df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


In [46]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=I5IGOLJKRUIUA1E4PMYEN40ZAO15GIVS33AUZS1HLDTJWLYS&client_secret=DFPEVUHB31554JZI5QDDRVGU5531ANMTCZSEDNOYTM51AZHG&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [47]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f1236e10092d14e7d22d8db'},
 'response': {'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.751976046055574,
          'lng': -79.33214044722958}],
        'distance': 245,
        'cc': 'CA'

In [48]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [49]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:6: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  


name         categories        lat        lng
0             Brookbanks Park               Park  43.751976 -79.332140
1              TTC stop #8380           Bus Stop  43.752672 -79.326351
2               Variety Store  Food & Drink Shop  43.751974 -79.333114
3  TTC stop - 44 Valley Woods           Bus Stop  43.755402 -79.333741

## Explore Neighborhoods in Toronto

In [51]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [53]:
#Get all the venues in each of the neighbourhoods in toronto

In [52]:
venues = getNearbyVenues(names=df['Neighborhood'],
                         latitudes=df['Latitude'],
                         longitudes=df['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [55]:
#get the size of venues
print(venues.shape)
venues.head()

(2163, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3         Parkwoods              43.753259              -79.329656   
4  Victoria Village              43.725882              -79.315572   

                        Venue  Venue Latitude  Venue Longitude  \
0             Brookbanks Park       43.751976       -79.332140   
1              TTC stop #8380       43.752672       -79.326351   
2               Variety Store       43.751974       -79.333114   
3  TTC stop - 44 Valley Woods       43.755402       -79.333741   
4      Victoria Village Arena       43.723481       -79.315635   

      Venue Category  
0               Park  
1           Bus Stop  
2  Food & Drink Shop  
3           Bus Stop  
4       Hockey Arena

In [57]:
#count the number of venues in each neighbourhood
venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                               7   
Bathurst Manor, Wilson Heights, Downsview North                     23   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   28   
...                                                                ...   
Willowdale, Willowdale East                                         33   
Willowdale, Willowdale West                                          7   
Woburn                                                               4   
Woodbine Heights                                                     7   
York Mills West                                                      3   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      23   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    28   
...                                                                 ...   
Willowdale, Willowdale East                                          33   
Willowdale, Willowdale West                                           7   
Woburn                                                                4   
Woodbine Heights                                                      7   
York Mills West                                                       3   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                               7               7   
Bathurst Manor, Wilson Heights, Downsview North     23              23   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   28              28   
...                                                ...             ...   
Willowdale, Willowdale East                         33              33   
Willowdale, Willowdale West                          7               7   
Woburn                                               4               4   
Woodbine Heights                                     7               7   
York Mills West                                      3               3   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                         7   
Bathurst Manor, Wilson Heights, Downsview North               23   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             28   
...                                                          ...   
Willowdale, Willowdale East                                   33   
Willowdale, Willowdale West                                    7   
Woburn                                                         4   
Woodbine Heights                                               7   
York Mills West                                                3   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                        7 

In [58]:
#count the number of unique venue categories
print('There are {} uniques categories.'.format(len(venues['Venue Category'].unique())))

There are 274 uniques categories.


## Analyze Each Neighborhood

In [60]:
# one hot encoding
tor_onehot = pd.get_dummies(venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
tor_onehot['Neighborhood'] = venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [tor_onehot.columns[-1]] + list(tor_onehot.columns[:-1])
tor_onehot = tor_onehot[fixed_columns]

Yoga Studio  Accessories Store  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Toy / Game Store  Trail  Train Station  \
0             0         0  ...                 0      0              0   
1             0         0  ...                 0      0              0   
2             0         0  ...                 0      0              0   
3             0         0  ...                 0      0              0   
4             0         0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wings Joint  Women's Store  
0                0         0            0              0  
1                0         0            0              0  
2                0         0            0              0  
3                0         0            0              0  
4                0         0            0              0  

[5 rows x 274 columns]

In [70]:
tor_onehot.shape

(2163, 274)

we got 274 columns for each of the 2163 venues in the neighbourhoods

In [68]:
# Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
tor_grouped = tor_onehot.groupby('Neighborhood').mean().reset_index()
tor_grouped

Neighborhood  Yoga Studio  \
0                                         Agincourt          0.0   
1                            Alderwood, Long Branch          0.0   
2   Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                   Bayview Village          0.0   
4                 Bedford Park, Lawrence Manor East          0.0   
..                                              ...          ...   
90                      Willowdale, Willowdale East          0.0   
91                      Willowdale, Willowdale West          0.0   
92                                           Woburn          0.0   
93                                 Woodbine Heights          0.0   
94                                  York Mills West          0.0   

    Accessories Store  Airport  Airport Food Court  Airport Lounge  \
0                 0.0      0.0                 0.0             0.0   
1                 0.0      0.0                 0.0             0.0   
2                 0.0      0.0                 0.0             0.0   
3                 0.0      0.0                 0.0             0.0   
4                 0.0      0.0                 0.0             0.0   
..                ...      ...                 ...             ...   
90                0.0      0.0                 0.0             0.0   
91                0.0      0.0                 0.0             0.0   
92                0.0      0.0                 0.0             0.0   
93                0.0      0.0                 0.0             0.0   
94                0.0      0.0                 0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0               0.0             0.000000           0.0  ...   
1               0.0               0.0             0.000000           0.0  ...   
2               0.0               0.0             0.000000           0.0  ...   
3               0.0               0.0             0.000000           0.0  ...   
4               0.0               0.0             0.035714           0.0  ...   
..              ...               ...                  ...           ...  ...   
90              0.0               0.0             0.000000           0.0  ...   
91              0.0               0.0             0.000000           0.0  ...   
92              0.0               0.0             0.000000           0.0  ...   
93              0.0               0.0             0.000000           0.0  ...   
94              0.0               0.0             0.000000           0.0  ...   

    Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0           0.000000    0.0            0.0                            0.0   
1           0.000000    0.0            0.0                            0.0   
2           0.000000    0.0            0.0                            0.0   
3           0.000000    0.0            0.0                            0.0   
4           0.035714    0.0            0.0                            0.0   
..               ...    ...            ...                            ...   
90          0.000000    0.0            0.0                            0.0   
91          0.000000    0.0            0.0                            0.0   
92          0.000000    0.0            0.0                            0.0   
93          0.000000    0.0            0.0                            0.0   
94          0.000000    0.0            0.0                            0.0   

    Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0                0.0               0.000000              0.0       0.0   
1                0.0               0.000000              0.0       0.0   
2                0.0               0.000000              0.0       0.0   
3                0.0               0.000000              0.0       0.0   
4                0.0               0.000000              0.0       0.0   
..               ...                   

In [72]:
num_top_venues = 5

for hood in tor_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = tor_grouped[tor_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0                     Lounge  0.25
1             Breakfast Spot  0.25
2  Latin American Restaurant  0.25
3               Skating Rink  0.25
4              Metro Station  0.00


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place  0.29
1  Sandwich Place  0.14
2     Coffee Shop  0.14
3             Pub  0.14
4        Pharmacy  0.14


----Bathurst Manor, Wilson Heights, Downsview North----
                       venue  freq
0                Coffee Shop  0.09
1                       Bank  0.09
2                Pizza Place  0.04
3  Middle Eastern Restaurant  0.04
4           Sushi Restaurant  0.04


----Bayview Village----
                 venue  freq
0                 Bank  0.25
1  Japanese Restaurant  0.25
2   Chinese Restaurant  0.25
3                 Café  0.25
4  Moroccan Restaurant  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0          Restaurant  0.07
1  Italian Restaurant  0.07

#### Let's put that into a *pandas* dataframe

In [141]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [191]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = tor_grouped['Neighborhood']

for ind in np.arange(tor_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tor_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant          Skating Rink   
1                   Gym                   Pharmacy        Sandwich Place   
2                  Bank           Sushi Restaurant    Frozen Yogurt Shop   
3                  Bank         Chinese Restaurant   Japanese Restaurant   
4             Juice Bar                 Restaurant    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Dumpling Restaurant   Distribution Center               Dog Run   
1           Coffee Shop                   Pub         Women's Store   
2        Sandwich Place                 Diner            Restaurant   
3         Women's Store   Distribution Center               Dog Run   
4           Coffee Shop              Boutique              Pharmacy   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                 Donut Shop              Drugstore  
1               Dog Run         Dim Sum Restaurant                  Diner  
2         Deli / Bodega  Middle Eastern Restaurant      Mobile Phone Shop  
3      Doner Restaurant                 Donut Shop              Drugstore  
4             Locksmith    Comfort Food Restaurant   Fast Food Restaurant

## Cluster Neighborhoods

In [192]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

tor_grouped_clustering = tor_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tor_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [193]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster_Labels', kmeans.labels_)

tor_merged = df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
tor_merged = tor_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

tor_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster_Labels 1st Most Common Venue  \
0  43.753259 -79.329656             4.0              Bus Stop   
1  43.725882 -79.315572             1.0          Hockey Arena   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                    Park     Food & Drink Shop        Discount Store   
1   Portuguese Restaurant           Pizza Place           Coffee Shop   
2                    Park                Bakery        Breakfast Spot   
3  Furniture / Home Store         Women's Store              Boutique   
4       College Cafeteria                 Diner           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   Distribution Center               Dog Run      Doner Restaurant   
1                 Diner        Discount Store   Distribution Center   
2                  Café                   Pub               Theater   
3    Miscellaneous Shop           Coffee Shop           Event Space   
4                   Bar              Beer Bar         Smoothie Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Donut Shop             Drugstore     Dim Sum Restaurant  
1                Dog Run      Doner Restaurant             Donut Shop  
2            Yoga Studio        Chocolate Shop      Electronics Store  
3  Vietnamese Restaurant     Accessories Store   Ethiopian Restaurant  
4         Sandwich Place         Burrito Place                   Café

In [194]:
tor_merged

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
0                                            Parkwoods  43.753259 -79.329656   
1                                     Victoria Village  43.725882 -79.315572   
2                            Regent Park, Harbourfront  43.654260 -79.360636   
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
..                                                 ...        ...        ...   
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944   
99                                Church and Wellesley  43.665860 -79.383160   
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509   
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999   

     Cluster_Labels 1st Most Common Venue       2nd Most Common Venue  \
0               4.0              Bus Stop                        Park   
1               1.0          Hockey Arena       Portuguese Restaurant   
2               1.0           Coffee Shop                        Park   
3               1.0        Clothing Store      Furniture / Home Store   
4               1.0           Coffee Shop           College Cafeteria   
..              ...                   ...                         ...   
98              1.0                  Pool                  Smoke Shop   
99              1.0           Coffee Shop         Japanese Restaurant   
100             1.0    Light Rail Station                      Garden   
101             4.0                  Park  Construction & Landscaping   
102             1.0         Grocery Store              Sandwich Place   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0       Food & Drink Shop        Discount Store   Distribution Center   
1             Pizza Place           Coffee Shop                 Diner   
2                  Bakery        Breakfast Spot                  Café   
3           Women's Store              Boutique    Miscellaneous Shop   
4                   Diner           Yoga Studio                   Bar   
..                    ...                   ...                   ...   
98                  River            Donut Shop    Dim Sum Restaurant   
99       Sushi Restaurant               Gay Bar            Restaurant   
100            Comic Shop           Pizza Place      Recording Studio   
101        Baseball Field             Drugstore        Discount Store   
102                Bakery       Supplement Shop           Flower Shop   

    6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0                 Dog Run      Doner Restaurant             Donut Shop   
1          Discount Store   Distribution Center                Dog Run   
2                     Pub               Theater            Yoga Studio   
3             Coffee Shop           Event Space  Vietnamese Restaurant   
4                Beer Bar         Smoothie Shop         Sandwich Place   
..                    ...                   ...                    ...   
98                  Diner        Discount Store    Distribution Center   
99            Yoga Studio                   Pub            Men's Store   
100            Restaurant         Burrito Place                Brewery   
101   Distribution Center               Dog Run       Doner Restaurant   
102         Tanning Salon     

In [195]:
tor_merged['Cluster_Labels'].unique()

array([ 4.,  1., nan,  3.,  0.,  2.])

In [196]:
tor_merged = tor_merged[tor_merged['Cluster_Labels'].notna()]

In [197]:
tor_merged.shape

(99, 16)

In [198]:
tor_merged.Cluster_Labels = tor_merged.Cluster_Labels.astype(int)

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/pandas/core/generic.py:5303: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [199]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [206]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Neighborhood'], tor_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [201]:
tor_merged.loc[tor_merged['Cluster_Labels'] == 0, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke               0   Filipino Restaurant         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11             Drugstore                 Diner        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11   Distribution Center               Dog Run      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
11            Donut Shop    Dumpling Restaurant

In [202]:
tor_merged.loc[tor_merged['Cluster_Labels'] == 1, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue  \
1          North York               1          Hockey Arena   
2    Downtown Toronto               1           Coffee Shop   
3          North York               1        Clothing Store   
4    Downtown Toronto               1           Coffee Shop   
7          North York               1                   Gym   
..                ...             ...                   ...   
97   Downtown Toronto               1           Coffee Shop   
98          Etobicoke               1                  Pool   
99   Downtown Toronto               1           Coffee Shop   
100      East Toronto               1    Light Rail Station   
102         Etobicoke               1         Grocery Store   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1     Portuguese Restaurant           Pizza Place           Coffee Shop   
2                      Park                Bakery        Breakfast Spot   
3    Furniture / Home Store         Women's Store              Boutique   
4         College Cafeteria                 Diner           Yoga Studio   
7                Beer Store   Japanese Restaurant            Restaurant   
..                      ...                   ...                   ...   
97                     Café                 Hotel                   Gym   
98               Smoke Shop                 River            Donut Shop   
99      Japanese Restaurant      Sushi Restaurant               Gay Bar   
100                  Garden            Comic Shop           Pizza Place   
102          Sandwich Place                Bakery       Supplement Shop   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1                   Diner        Discount Store   Distribution Center   
2                    Café                   Pub               Theater   
3      Miscellaneous Shop           Coffee Shop           Event Space   
4                     Bar              Beer Bar         Smoothie Shop   
7                    Café           Coffee Shop      Asian Restaurant   
..                    ...                   ...                   ...   
97             Restaurant   Japanese Restaurant    Seafood Restaurant   
98     Dim Sum Restaurant                 Diner        Discount Store   
99             Restaurant           Yoga Studio                   Pub   
100      Recording Studio            Restaurant         Burrito Place   
102           Flower Shop         Tanning Salon           Social Club   

     8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
1                  Dog Run          Doner Restaurant             Donut Shop  
2              Yoga Studio            Chocolate Shop      Electronics Store  
3    Vietnamese Restaurant         Accessories Store   Ethiopian Restaurant  
4           Sandwich Place             Burrito Place                   Café  
7           Discount Store            Sandwich Place              Bike Shop  
..                     ...                       ...                    ...  
97        Asian Restaurant             Deli / Bodega    American Restaurant  
98     Distribution Center                   Dog Run       Doner Restaurant  
99             Men's Store  Mediterranean Restaurant                  Hotel  
100                Brewery                Skate Park         Farmers Market  
102             Kids Store    Thrift / Vintage Store           Burger Joint  

[85 rows x 12 columns]

In [203]:
tor_merged.loc[tor_merged['Cluster_Labels'] == 2, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue  \
62  Central Toronto               2                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62            Donut Shop    Dim Sum Restaurant                 Diner   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
62        Discount Store   Distribution Center               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
62      Doner Restaurant         Women's Store           Dessert Shop

In [204]:
tor_merged.loc[tor_merged['Cluster_Labels'] == 3, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough               3  Fast Food Restaurant             Drugstore   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6                 Diner        Discount Store   Distribution Center   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6               Dog Run      Doner Restaurant            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
6   Dumpling Restaurant       Hakka Restaurant

In [205]:
tor_merged.loc[tor_merged['Cluster_Labels'] == 4, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue  \
0          North York               4              Bus Stop   
21               York               4                  Park   
35          East York               4                  Park   
49         North York               4                Bakery   
61    Central Toronto               4                  Park   
64               York               4                  Park   
66         North York               4                  Park   
68    Central Toronto               4                  Park   
85        Scarborough               4                  Park   
91   Downtown Toronto               4                  Park   
101         Etobicoke               4                  Park   

          2nd Most Common Venue       3rd Most Common Venue  \
0                          Park           Food & Drink Shop   
21                         Pool               Women's Store   
35            Convenience Store                 Coffee Shop   
49                         Park  Construction & Landscaping   
61                  Swim School                    Bus Line   
64            Convenience Store         Dumpling Restaurant   
66            Convenience Store           Electronics Store   
68             Sushi Restaurant                       Trail   
85                       Bakery                  Playground   
91                   Playground                       Trail   
101  Construction & Landscaping              Baseball Field   

    4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0          Discount Store   Distribution Center               Dog Run   
21       Greek Restaurant         Deli / Bodega  Ethiopian Restaurant   
35              Drugstore                 Diner        Discount Store   
49                  Trail      Basketball Court             Drugstore   
61             Donut Shop        Discount Store   Distribution Center   
64         Discount Store   Distribution Center               Dog Run   
66              Drugstore        Discount Store   Distribution Center   
68          Jewelry Store         Women's Store            Donut Shop   
85             Donut Shop                 Diner        Discount Store   
91             Donut Shop    Dim Sum Restaurant                 Diner   
101             Drugstore        Discount Store   Distribution Center   

    7th Most Common Venue        8th Most Common Venue 9th Most Common Venue  \
0        Doner Restaurant                   Donut Shop             Drugstore   
21      Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
35    Distribution Center                      Dog Run      Doner Restaurant   
49         Discount Store          Distribution Center               Dog Run   
61                Dog Run             Doner Restaurant             Drugstore   
64       Doner Restaurant                   Donut Shop             Drugstore   
66                Dog Run             Doner Restaurant            Donut Shop   
68                  Diner               Discount Store   Distribution Center   
85    Distribution Center                      Dog Run      Doner Restaurant   
91         Discount Store          Distribution Center               Dog Run   
101               Dog Run             Doner Restaurant            Donut Shop   

          10th Most Common Venue  
0             Dim Sum Restaurant  
21                     Drugstore  
35                    Donut Shop  
49              Doner Restaurant  
61            Dim Sum Restaurant  
64   Eastern European Restaurant  
66   Eastern European Restaurant  
68                       Dog Run  
85                     Drugstore  
91              Doner Restaurant  
101          Dumpling Restaurant